### PR TITLE
Enable FileStream length caching with FileShare.Delete

### PR DESCRIPTION
**Pull request by pentp**
`FileShare.Delete` doesn't allow file length to change.
Also cleaned up the `share` parameter from handle based strategy constructors.

### CODE REVIEW

**Comment by stephentoub** [comment]
> Enable FileStream length caching with FileShare.Delete

I must be missing it, but I don't see any changes related to FileShare.Delete...?

**Comment by pentp** [comment]
Previously the check was `_share <= FileShare.Read`, now it's `(share & FileShare.Write) == 0` so it allows length caching if `FileShare.Read | FileShare.Delete` is used to open a file.

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs:20]: It seems that we should change the name of this field. Historically it meant "the handle has been exposed, validate file offset on every async operaiton":

https://github.com/dotnet/runtime/blob/50413eafd5a930dc5f3cce1a7fcba743cbb36c80/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.cs#L301-L303

Now it means "don't cache file length as it's not safe".

```suggestion
        private bool _lengthCanBeCached;
```

**Review by adamsitnik — COMMENTED** [review]
Overall LGTM, but it would be great to rename the field before we hit the merge button.

@pentp thank you!

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @pentp !

**Pull request merged**
